blog/2020/12/package_latency.md

```diff
@@ -0,0 +1,718 @@
+                          cs_vec::Vector{Any},
+                          keepkeys::Bool, ungroup::Bool, copycols::Bool,
+                          keeprows::Bool, renamecols::Bool)
+    if any(x -> x isa Pair && first(x) isa Tuple, cs_vec)
+...
+```
+
+eliminates most of the inference time.
+
+These and a few other similar changes reduced `@time include("grouping.jl")` in a fresh session by nearly 15%, from 283s to 248s.  In general, reducing argument type-diversity in long methods is one of the easiest ways to reduce latency.  Here we've handled this just by splitting methods; if we had a deeper understanding of the package internals, we might consider more extensive refactoring that reduces the diversity at an earlier stage of the call chain and yields even greater benefits.  In either case, the new `@snoopi_deep` makes it easy to discover which methods would benefit most from this kind of treatment.
+
+## Strategy #2: inspecting and addressing inference failures
+
+Some calls produce a result with unpredictable type, and thus to an inference "failure."  Such failures can be roughly divided into "avoidable" and "unavoidable" failures.  An example of an "unavoidable" failure might be `data = load("myfile.ext")`, where the type of `data` depends on the contents of `"myfile.ext"`--there is no plausible way for inference to know what type to expect of `data`.  In cases like these, it is expected that any calls made to process `data` will have to be made by *runtime dispatch*, in which execution is paused to ask `data` for its type and then search the method tables for compatible methods.
+
+In constrast, some types of inference failure are avoidable. For example, in
+```julia
+s = []
+push!(s, 1, 2)
+sum(s)
+```
+an inference failure in the `sum` could have been avoided by defining `s = Int[]` rather than `s = []`.
+
+In real-world circumstances, you're likely to encounter a mixture of avoidable and unavoidable failures.  Fortunately, it will turn out to be possible to make progress with both types.
+
+### Identifying inference failures
+
+In addition to its ability to measure inference time for each method, the new `@snoopi_deep` also introduces a new capability that makes identifying inference failures easier than ever: while snooping, inference collects a [backtrace] upon each new (runtime) entrance into inference.  This makes it possible to discover the caller that triggered runtime dispatch, thus identifying the method and call that fails to infer.  When a package is well-inferred, most of these entrances come from toplevel (the commands you issue in the REPL, or inside a `@testset`), but by default these are excluded from SnoopCompile's reports; the goal is to identify internal, hopefully-avoidable inference failures.
+
+Let's apply this to DataFrames. After collecting the data with `@snoopi_deep include("runtests.jl")`, we can see inference failures with
+
+```julia
+julia> ibs = SnoopCompile.inference_breaks(tinf)
```

Comment:
I can't find `inference_breaks` in SnoopCompile (latest master), did the name change?